pkg/kv/kvserver/replica_raftlog_test.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kvserver
+
+import (
+	"context"
+	"math"
+	"math/rand/v2"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/raftentry"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplicaLogStorage(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	rt := makeReplicaLogStorageTest(t)
+	rt.start(ctx)
+	defer rt.stop(ctx)
+
+	seed := rand.Uint64()
+	t.Logf("seed: %d", seed)
+	rng := newRNG(seed)
+
+	// Start separate threads for truncations and reads under mu. Use stable seed
+	// for each thread.
+	// NB: stable seeds do not achieve reproducibility here, but we try hard. Note
+	// that at least the append thread below is reproducible.
+	for i := 0; i < 5; i++ {
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "read"},
+			func(ctx context.Context) { rt.readLoop(ctx, fork) }))
```

Comment:
These readers hold `Replica.mu`. Might as well add a `raftMu`-based reader, since we now mostly use the log under `raftMu`. It's less interesting to test here because it won't be concurrent with the `writeLoop`.